### PR TITLE
New module: Manage Foreman and Katello entities

### PR DESCRIPTION
**Pull request by ehelms**
##### ISSUE TYPE
- New Module Pull Request
##### COMPONENT NAME
- foreman/foreman.py
- foreman/katello.py
##### ANSIBLE VERSION

```
ansible 2.1.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```
##### SUMMARY

This add some initial entity management for Katello [1] which is a plugin to Foreman [2]. Given that Foreman has a core with plugin architecture I have chosen to break these into separate files representing each code base such that the entities are managed respective to the originating code base. The idea is to allow users to manage their content inventory within the server via Ansible so more easily recreate, add to or alter their setups. This makes automating the process of rebuilding, migrating or populating a new Foreman organization with data easier.

I did not feel any of the existing categories properly captured the domain that Foreman and Katello are in and chose to create a new category. If this is not allowed, please offer some suggestions of where this would be better off living.

[1] http://www.katello.org/
[2] https://theforeman.org/

### CODE REVIEW

**Comment by gregdek** [comment]
Thanks @ehelms for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
Thanks @ehelms for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by lpramuk** [comment]
Python2.4 platform is not among Foreman and Katello platforms.
Failed python24 can be waived then?

**Comment by gundalow** [comment]
needs_revision

You will need to edit `.travis.yml` and update the `-x` (exclude) line approx line 120 to ignore this new directory.

File: management/foreman/foreman.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+---
+module: foreman
+short_description: Manage Foreman Resources
+description:
+    - Allows the management of Foreman resources inside your Foreman server
+version_added: "2.2"
+author: "Eric D Helms (@ehelms)"
+requirements:
+    - nailgun python module
+options:
```
Comment (gundalow) [inline management/foreman/foreman.py]: Missing `entity` and `params` documentation

File: management/foreman/katello.py [hunk 2]
```diff
@@ -0,0 +1,444 @@
...
+---
+module: katello
+short_description: Manage Katello Resources
+description:
+    - Allows the management of Katello resources inside your Foreman server
+version_added: "2.2"
+author: "Eric D Helms (@ehelms)"
+requirements:
+    - nailgun python module
+options:
```
Comment (gundalow) [inline management/foreman/katello.py]: Missing `entity` and `params` documentation

**Comment by gundalow** [comment]
Few minor docs issues. I don't have access to foreman so can't test, code looks sensible

**Comment by gundalow** [comment]
The view from The Core Team is that "management" is too broad,  and could apply to anything, what do you think of "web_infrastructure"?

**Comment by ehelms** [comment]
@gundalow If you'd like to give this a try with Foreman/Katello we provide a Vagrant based setup to easily spin up an installation (https://github.com/Katello/forklift). For the categorization question, I'm not sure 'web_infrastructure' is that discoverable for what it is compared to say the cloud category which is more obvious. Given management is too broad, some other ideas that come to mind are 'infrastructure_management', and 'systems_management'.

**Comment by ehelms** [comment]
Updated documentation

**Comment by ehelms** [comment]
@gundalow are there any updates around the previous comments? I have rebased this to the latest devel. I would like to get this in in order to steer further development here and in some hope of pushing all of the Satellite modules to this upstream focused module.

**Comment by ehelms** [comment]
I've updated this to introduce a new category after some further thought of 'infrastructure'. I think this may better capture the category of where Foreman/Katello live, is broad in scope, and would allow other infrastructure management projects a place to live. I don't think any current category captures this properly and I'd prefer 'cloud' not become a dumping ground for everything that has some cloud integration as it works well as a place for cloud providers.

@gundalow would this work? How do I get this moving forward so that it does not become stagnant and more contributions can be made in this space?

File: .travis.yml [hunk 3]
```diff
@@ -117,7 +117,7 @@ install:
   - pip install git+https://github.com/ansible/ansible.git@devel#egg=ansible
   - pip install git+https://github.com/sivel/ansible-testing.git#egg=ansible_testing
 script:
-  - python2.4 -m compileall -fq -x 'cloud/|monitoring/zabbix.*\.py|/dnf\.py|/layman\.py|/maven_artifact\.py|clustering/(consul.*|znode)\.py|notification/pushbullet\.py|database/influxdb/influxdb.*\.py|database/mssql/mssql_db\.py|/letsencrypt\.py|network/f5/bigip.*\.py|bmc/ipmi/.*\.py' .
+  - python2.4 -m compileall -fq -x 'cloud/|monitoring/zabbix.*\.py|/dnf\.py|/layman\.py|/maven_artifact\.py|clustering/(consul.*|znode)\.py|notification/pushbullet\.py|database/influxdb/influxdb.*\.py|database/mssql/mssql_db\.py|/letsencrypt\.py|network/f5/bigip.*\.py|bmc/ipmi/.*\.py|management' .
```
Comment (gundalow) [inline .travis.yml]: Needs updating to be `infrastructure/foreman`, this will fix the Travis issue

File: infrastructure/foreman/foreman.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+DOCUMENTATION = '''
+---
+module: foreman
+short_description: Manage Foreman Resources
+description:
+    - Allows the management of Foreman resources inside your Foreman server
+version_added: "2.2"
+author: "Eric D Helms (@ehelms)"
+requirements:
+    - nailgun python module
```
Comment (gundalow) [inline infrastructure/foreman/foreman.py]: - Python version as you've said 2.4 isn't supported 

Same fix on other modules

**Comment by gundalow** [comment]
Hi,
Apologies for the delay, this seems to have gotten lost in the time between holiday and AnsibleFest.

I'm happy with `infrastructure/foreman`

needs_revision due to a few docs & Travis fixes

**Comment by ehelms** [comment]
@gundalow thanks for the re-review. I've updated the requirements and travis file.

File: infrastructure/foreman/katello.py [hunk 5]
```diff
@@ -0,0 +1,454 @@
...
+---
+module: katello
+short_description: Manage Katello Resources
+description:
+    - Allows the management of Katello resources inside your Foreman server
+version_added: "2.2"
+author: "Eric D Helms (@ehelms)"
+requirements:
+    - nailgun
+    - module
```
Comment (gundalow) [inline infrastructure/foreman/katello.py]: ditto

File: infrastructure/foreman/foreman.py [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+---
+module: foreman
+short_description: Manage Foreman Resources
+description:
+    - Allows the management of Foreman resources inside your Foreman server
+version_added: "2.2"
+author: "Eric D Helms (@ehelms)"
+requirements:
+    - nailgun
+    - module
```
Comment (gundalow) [inline infrastructure/foreman/foreman.py]: Should this be part of the line above?

File: infrastructure/foreman/foreman.py [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+        return True
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            server_url=dict(required=True),
+            username=dict(required=True, no_log=True),
+            password=dict(required=True, no_log=True),
+            entity=dict(required=True, no_log=False),
+            verify_ssl=dict(required=False, type='bool', default=False),
```
Comment (gundalow) [inline infrastructure/foreman/foreman.py:114]: `verify_ssl` isn't in the docs, please add it.

File: infrastructure/foreman/katello.py [hunk 8]
```diff
@@ -0,0 +1,454 @@
...
+        return True
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            server_url=dict(required=True),
+            username=dict(required=True, no_log=True),
+            password=dict(required=True, no_log=True),
+            entity=dict(required=True, no_log=False),
+            verify_ssl=dict(required=False, type='bool', default=False),
```
Comment (gundalow) [inline infrastructure/foreman/katello.py:460]: Not in the docs, you should be able to copy & paste from another module.

**Comment by gundalow** [comment]
```
username=dict(required=True, no_log=True),
password=dict(required=True, no_log=True),
entity=dict(required=True, no_log=False),
```

It's right not to log passwords, though should we be logging `username` and `entity`?

File: infrastructure/foreman/foreman.py (hunk 7 above)
Comment (ehelms) [inline infrastructure/foreman/foreman.py:114]: Sorry - I don't follow what you mean by copy/paste in this context.

Comment (ehelms) [inline infrastructure/foreman/foreman.py:114]: Oh nevermind me, I get the two comments together.

**Comment by ehelms** [comment]
Updated to include verify_ssl in the documentation, updated the requirements list to make more sense and set username to now be logged along with entity. I can see these being useful for auditing purposes.

**Comment by ehelms** [comment]
@gundalow I've updated and rebased the PR per your comments.

File: infrastructure/foreman/foreman.py [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+    - Allows the management of Foreman resources inside your Foreman server
+version_added: "2.2"
+author: "Eric D Helms (@ehelms)"
+requirements:
+    - nailgun
+    - "python >= 2.7"
+options:
+    server_url:
+        description:
+            - URL of Foreman server
```
Comment (gundalow) [inline infrastructure/foreman/foreman.py:35]: Please can you add fullstop `.` at the end of all the `descriptions:`

File: infrastructure/foreman/katello.py [hunk 10]
```diff
@@ -0,0 +1,458 @@
...
+            activation_key.environment = lifecycle_environment
+            activation_key.update()
+
+        return True
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            server_url=dict(required=True),
+            username=dict(required=True, no_log=False),
```
Comment (gundalow) [inline infrastructure/foreman/katello.py]: You can delete
`no_log=False` as it's the default

File: infrastructure/foreman/katello.py [hunk 11]
```diff
@@ -0,0 +1,458 @@
...
+            password=dict(required=True, no_log=True),
+            entity=dict(required=True, no_log=False),
```
Comment (gundalow) [inline infrastructure/foreman/katello.py:458]: You can delete
`no_log=False` as it's the default

File: infrastructure/foreman/katello.py [hunk 12]
```diff
@@ -0,0 +1,458 @@
...
+            verify_ssl=dict(required=False, type='bool', default=False),
```
Comment (gundalow) [inline infrastructure/foreman/katello.py:460]: You can delete
`no_log=False` as it's the default

**Comment by gundalow** [comment]
I'm guessing the merge conflict is due to recent changes to `.travis.yml`

**Comment by ehelms** [comment]
@gundalow thanks for the reviews, PR is updated and green

**Comment by gundalow** [comment]
Shipit

File: infrastructure/foreman/foreman.py [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+        self._module = module
+
+    def find_organization(self, name, **params):
+        org = self._entities.Organization(self._server, name=name, **params)
+        response = org.search(set(), {'search': 'name={}'.format(name)})
+
+        if len(response) == 1:
+            return response[0]
+        else:
+            self._module.fail_json(msg="No Content View found for %s" % name)
```
Comment (stbenjam) [inline infrastructure/foreman/foreman.py:91]: Should this really say "Content view" here?

**Comment by stbenjam** [comment]
@ehelms Trying this out. Here's what's in the role:

``` yaml
---
- name: "Create Organization"
  action:
    module: foreman
    username: "admin"
    password: "changeme"
    server_url: "https://centos7-katello-p4-nightly.example.com"
    entity: "organization"
       params:
       name: "My Cool New Organization"
```

It fails with:

```
TASK [test : Create Organization] **********************************************
fatal: [centos7-katello-p4-nightly]: FAILED! => {"changed": false, "failed": true, "msg": "No Content View found for ********"}
```

In production.log

```
2016-09-14 09:55:07 cf1326e2 [app] [I] Processing by Katello::Api::V2::OrganizationsController#index as */*
2016-09-14 09:55:07 cf1326e2 [app] [I]   Parameters: {"search"=>"name=My Cool New Organization", "api_version"=>"v2", "organization"=>{}}
2016-09-14 09:55:07 cf1326e2 [app] [I] Authorized user admin(Admin User)
2016-09-14 09:55:07 cf1326e2 [app] [I]   Rendered /opt/theforeman/tfm/root/usr/share/gems/gems/katello-3.2.0/app/views/katello/api/v2/organizations/index.json.rabl within api/v2/layouts/index_layout (1.1ms)
2016-09-14 09:55:07 cf1326e2 [app] [I] Completed 200 OK in 24ms (Views: 2.2ms | ActiveRecord: 10.4ms)
```

It searchs for the org and doesn't find one.  Shouldn't it create it in that case?

File: infrastructure/foreman/foreman.py (hunk 13 above)
Comment (ehelms) [reply infrastructure/foreman/foreman.py:91]: I think we generally refer to it as a 'Content View' being an entity

Comment (ehelms) [reply infrastructure/foreman/foreman.py:91]: Oh actually this should just say "No organization found", however, I removed this helper method for now as this behavior doesn't need to be used presently. Thus I updated it to also fix the issue you found below to create an org.

**Comment by ehelms** [comment]
@gundalow what is the process for this getting merged? I saw you said 'Shipit' but then it sat for a few days.

**Comment by gundalow** [comment]
@stbenjam do the updates to this PR now work for you?

**Comment by gundalow** [comment]
@ehelms Hi. New modules require two ship its. I gave this the thumbs up, however stbenjam has reported an issue about.

Unfortunately this has now hit feature freeze in preparation for the 2.2 release. Once we've created the branch for the first RC (release candidate) the `devel` branch will be opened up again and the module can be merged so it will be released in 2.3.

If people want to use the his module without having to run from devel they can copy the file into a `./library` directory which is at the same level as `roles`.

**Comment by stbenjam** [comment]
@gundalow Yes, it works for me now:

```
TASK [test : Create Organization] **********************************************
task path: /home/stbenjam/git/forklift/playbooks/roles/test/tasks/main.yml:2
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: stbenjam
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1474659716.01-253399595976826 `" && echo ansible-tmp-1474659716.01-253399595976826="` echo $HOME/.ansible/tmp/ansible-tmp-1474659716.01-253399595976826 `" ) && sleep 0'
<localhost> PUT /tmp/tmpKk394r TO /home/stbenjam/.ansible/tmp/ansible-tmp-1474659716.01-253399595976826/foreman
<localhost> EXEC /bin/sh -c 'sudo -H -S -n -u root /bin/sh -c '"'"'echo BECOME-SUCCESS-oyfaeppfpzlmnabjmgvgqiuxalecceff; LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /home/stbenjam/.ansible/tmp/ansible-tmp-1474659716.01-253399595976826/foreman; rm -rf "/home/stbenjam/.ansible/tmp/ansible-tmp-1474659716.01-253399595976826/" > /dev/null 2>&1'"'"' && sleep 0'
changed: [centos7-provision-nightly -> localhost] => {"changed": true, "invocation": {"module_args": {"entity": "organization", "params": {}, "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", "server_url": "https://centos7-provision-nightly.example.com", "username": "admin", "verify_ssl": false}, "module_name": "foreman"}, "result": "organization updated"}

PLAY RECAP *********************************************************************
centos7-provision-nightly  : ok=2    changed=1    unreachable=0    failed=0   
```

Shipit

**Comment by ehelms** [comment]
@gundalow Hey, looks like 2.2 has been branched and I've updated this PR to pass tests. Given the previous two shipits what is needed to merge this?

**Comment by gundalow** [comment]
Merged

**Comment by ehelms** [comment]
Woot! Thanks!

**Comment by gundalow** [comment]
You are most welcome,  thanks for the contribution. We look forward to more PRs and seeing you around the meetings on IRC #ansible-meeting on Freenode.

**Pull request merged**